pkg/restore/change_storageclass_action.go

```diff
@@ -87,33 +91,73 @@ func (a *ChangeStorageClassAction) Execute(input *velero.RestoreItemActionExecut
 		"name":      obj.GetName(),
 	})
 
-	// use the unstructured helpers here since this code is for both PVs and PVCs, and the
-	// field names are the same for both types.
-	storageClass, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "storageClassName")
-	if err != nil {
-		return nil, errors.Wrap(err, "error getting item's spec.storageClassName")
+	// change StatefulSet volumeClaimTemplates storageClassName
+	if obj.GetKind() == "StatefulSet" {
+		sts := new(appsv1.StatefulSet)
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.UnstructuredContent(), sts); err != nil {
+			return nil, err
+		}
+
+		if len(sts.Spec.VolumeClaimTemplates) > 0 {
+			for index, pvc := range sts.Spec.VolumeClaimTemplates {
+				warn, err, newStorageClass := a.isStorageClassExist(*pvc.Spec.StorageClassName, config)
+				if warn != "" && err == nil {
+					log.Debug(warn)
+					continue
+				} else if warn == "" && err != nil {
+					return nil, err
+				}
+
+				log.Infof("Updating item's storage class name to %s", newStorageClass)
+				sts.Spec.VolumeClaimTemplates[index].Spec.StorageClassName = &newStorageClass
+			}
+
+			newObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(sts)
+			if err != nil {
+				return nil, errors.Wrap(err, "convert obj to StatefulSet failed")
+			}
+			obj.Object = newObj
+		}
+	} else {
+		// use the unstructured helpers here since this code is for both PVs and PVCs, and the
+		// field names are the same for both types.
+		storageClass, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "storageClassName")
+		if err != nil {
+			return nil, errors.Wrap(err, "error getting item's spec.storageClassName")
+		}
+
+		warn, err, newStorageClass := a.isStorageClassExist(storageClass, config)
+		if warn != "" && err == nil {
+			log.Debug(warn)
+			return velero.NewRestoreItemActionExecuteOutput(input.Item), nil
+		} else if warn == "" && err != nil {
+			return nil, err
+		}
+
+		log.Infof("Updating item's storage class name to %s", newStorageClass)
+
+		if err := unstructured.SetNestedField(obj.UnstructuredContent(), newStorageClass, "spec", "storageClassName"); err != nil {
+			return nil, errors.Wrap(err, "unable to set item's spec.storageClassName")
+		}
 	}
+
+	return velero.NewRestoreItemActionExecuteOutput(obj), nil
+}
+
+func (a *ChangeStorageClassAction) isStorageClassExist(storageClass string, cm *corev1.ConfigMap) (warn string, err error, newStorageClass string) {
```

Comment:
I'm a little confused about the function. Usually, if a function called `isXXXExist`, it returns a boolean value to indicate whether the resource exists or not.
Seems the returning `warn` here is only used to print some debug logs, how about logging the debug message inside the function directly and change the return values of the function to `(exists bool, newStorageClass string, err error)`?
If the function returns an error during the calling, returns the error directly. If an error is nil and exists is false, do continue, otherwise do the actual work

Reply:
Because I want `logSource` field to be on an optimal path when I need to print debug logs.
Your suggestion is a better way to deal with it, I will change the `isStorageClassExist` function.

Reply:
Because I want `logSource` field to be on an optimal path when I need to print debug logs.
Your suggestion is a better way to deal with it, I will change the `isStorageClassExist` function.